apisix/plugins/auth-casdoor.lua

```diff
@@ -0,0 +1,128 @@
+--
+-- Licensed to the Apache Software Foundation (ASF) under one or more
+-- contributor license agreements.  See the NOTICE file distributed with
+-- this work for additional information regarding copyright ownership.
+-- The ASF licenses this file to You under the Apache License, Version 2.0
+-- (the "License"); you may not use this file except in compliance with
+-- the License.  You may obtain a copy of the License at
+--
+--     http://www.apache.org/licenses/LICENSE-2.0
+--
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+--
+local core = require("apisix.core")
+local log = core.log
+local ngx = ngx
+local session = require("resty.session")
+local http = require("resty.http")
+local cjson = require("cjson")
+
+local plugin_name = "auth-casdoor"
+local schema = {
+    type = "object",
+    properties = {
+        casdoor_endpoint = { type = "string" },
+        casdoor_client_id = { type = "string" },
+        casdoor_client_secret = { type = "string" },
+        casdoor_callback_url = { type = "string" },
+    },
+    required = {
+        "casdoor_callback_url",
+        "casdoor_endpoint",
+        "casdoor_client_id",
+        "casdoor_client_secret"
+    }
+}
+
+local _M = {
+    version = 0.1,
+    priority = 12,
+    name = plugin_name,
+    schema = schema,
+}
+
+local function get_path(uri)
+    local without_query = uri:match("(.-)%?") or uri
+    return without_query:match(".-//[^/]+(/.*)") or without_query
+end
+
+local function fetch_access_token(conf)
+    local args = ngx.req.get_uri_args()
+    if not args.code or not args.state then
+        log.err("failed when accessing token. Invalid code or state ")
+        return nil
```

Comment:
```suggestion
        return nil, err
```
and log the err outside would be better